src/Kestrel.Core/KestrelConfigurationLoader.cs

```diff
@@ -259,6 +263,76 @@ private void LoadDefaultCert(ConfigurationReader configReader)
                     Options.DefaultCertificate = defaultCert;
                 }
             }
+            else
+            {
+                var logger = Options.ApplicationServices.GetRequiredService<ILogger<KestrelServer>>();
+                var certificate = FindDeveloperCertificateFile(configReader, logger);
+                if (certificate != null)
+                {
+                    logger.LocatedDevelopmentCertificate(certificate);
+                    Options.DefaultCertificate = certificate;
+                }
+            }
+        }
+
+        private X509Certificate2 FindDeveloperCertificateFile(ConfigurationReader configReader, ILogger<KestrelServer> logger)
+        {
+            string certificatePath = null;
+            try
+            {
+                if (configReader.Certificates.TryGetValue("Development", out var certificateConfig) &&
+                    certificateConfig.Path == null &&
+                    certificateConfig.Password != null &&
+                    TryGetCertificatePath(out certificatePath) &&
+                    File.Exists(certificatePath))
+                {
+                    var certificate = new X509Certificate2(certificatePath, certificateConfig.Password);
+                    return IsDevelopmentCertificate(certificate) ? certificate : null;
+                }
+                else if (!File.Exists(certificatePath))
+                {
+                    logger.FailedToLocateDevelopmentCertificateFile(certificatePath);
+                }
+            }
+            catch (CryptographicException)
+            {
+                logger.FailedToLoadDevelopmentCertificate(certificatePath);
+            }
+
+            return null;
+        }
+
+        private bool IsDevelopmentCertificate(X509Certificate2 certificate)
+        {
+            if (!string.Equals(certificate.Subject, "CN=localhost", StringComparison.Ordinal))
+            {
+                return false;
+            }
+
+            foreach (var ext in certificate.Extensions)
+            {
+                if (string.Equals(ext.Oid.Value, CertificateManager.AspNetHttpsOid, StringComparison.Ordinal))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private bool TryGetCertificatePath(out string path)
+        {
+            var hostingEnvironment = Options.ApplicationServices.GetRequiredService<IHostingEnvironment>();
+            var appName = hostingEnvironment.ApplicationName;
```

Comment:
The use of app name as the cert name still doesn't make sense. This is one more variable that can make this fail. Why can't you use a single instance named aspnetdevcert? Everything else is hard coded.

Reply:
Discussed offline